_includes/tutorials/message-ordering/kafka/code/src/main/java/io/confluent/developer/KafkaProducerApplication.java

```diff
@@ -0,0 +1,111 @@
+package io.confluent.developer;
+
+
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.serialization.StringSerializer;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+public class KafkaProducerApplication {
+
+    private final Producer<String, String> producer;
+    final String outTopic;
+
+    public KafkaProducerApplication(final Producer<String, String> producer,
+                                    final String topic) {
+        this.producer = producer;
+        outTopic = topic;
+    }
+
+    public Future<RecordMetadata> produce(final String message) {
+        final String[] parts = message.split("-");
+        final String key, value;
+        if (parts.length > 1) {
+            key = parts[0];
+            value = parts[1];
+        } else {
+            key = "NO-KEY";
```

Comment:
Maybe it's intentional, but `NO-KEY` _is_ a key, and will result in all records being sent to the same partition (vs `null`, which would result in using the uniform sticky partitioner instead)

Reply:
@mikebin good point.  Not intentional in that this code is duplicated from `kafka-producer-application` (cc: @bbejeck ) but in that case, IIUC the partitioner didn't matter.

So stepping back, what is the desirable behavior _for this tutorial_ if there is no key?  I think a true `null` and round robin to use both partitions would better demonstrate ordering on a per-partition basis.  If you agree, I'll modify the code accordingly.  WDYT?

Reply:
`null` makes sense if no key is provided, although in this specific tutorial, keys are provided the input data, so only if the user experiments with their own data would that code be relevant.

Reply:
I've tested this and it seems that the default partitioner behavior is, as described in https://github.com/apache/kafka/blob/2.6/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java#L30:

> If no partition or key is present choose the sticky partition that changes when the batch is full

Note that it *changes when the batch is full*, and all the null-keyed messages get put into the same batch so they all land in the same partition anyway.  If we wanted to force a different behavior, we could set `batch.size=0` to get more batches and get some kind of distribution, but that conflates different subjects in one KT and could lead to bad copy/paste.

So I'm thinking to change the code as follows:

```
diff --git a/_includes/tutorials/message-ordering/kafka/code/src/main/java/io/confluent/developer/KafkaProducerApplication.java b/_includes/tutorials/message-ordering/kafka/code/src/main/java/io/confluent/developer/KafkaProducerApplication.java
index 62a00c7c..45c9fb87 100644
--- a/_includes/tutorials/message-ordering/kafka/code/src/main/java/io/confluent/developer/KafkaProducerApplication.java
+++ b/_includes/tutorials/message-ordering/kafka/code/src/main/java/io/confluent/developer/KafkaProducerApplication.java
@@ -32,12 +32,12 @@ public class KafkaProducerApplication {
 
     public Future<RecordMetadata> produce(final String message) {
         final String[] parts = message.split("-");
-        final String key, value;
+        String key, value;
+        key = null;
         if (parts.length > 1) {
             key = parts[0];
             value = parts[1];
         } else {
-            key = "NO-KEY";
             value = parts[0];
         }
         final ProducerRecord<String, String> producerRecord = new ProducerRecord<>(outTopic, key, value);
```

...such that a real `null` key gets written, but we just need to understand that it doesn't necessarily mean they will round robin because they will likely land in the same batch.

Reply:
SGTM

Reply:
The "NO-KEY" was intentional to provide a default key if a record didn't have one after splitting the line.  Probably poorly worded.